src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -93,30 +152,162 @@ void GenericOTARequestorDriver::UpdateSuspended(System::Clock::Seconds32 delay)
     {
-        HandleError(state, error);
+        ChipLogProgress(SoftwareUpdate, "State is not kIdle, ignoring the AnnounceOTAProviders. State: %d",
+                        (int) mRequestor->GetCurrentUpdateState());
+        return;
     }
+
+    // Point the OTARequestor to the announced provider
+    mRequestor->SetCurrentProviderLocation(providerLocation);
+    mLastUsedProvider = providerLocation;
+
+    ScheduleDelayedAction(System::Clock::Seconds32(msToStart / 1000), StartDelayTimerHandler, this);
+}
+
+void GenericOTARequestorDriver::DriverSendQuery()
+{
+
+    // IMPLEMENTATION CHOICE
+    // In this implementation explicitly triggering a query cancels any in-progress update.
+    UpdateCancelled();
+
+    // Default provider timer only runs when there is no ongoing query/update; must stop it now.
+    // ConnectToProvider() will change the state from kIdle
+    StopDefaultProviderTimer();
+
+    mRequestor->ConnectToProvider(OTARequestorInterface::kQueryImage);
+}
+
+void GenericOTARequestorDriver::DefaultProviderTimerHandler(System::Layer * systemLayer, void * appState)
+{
+    ChipLogProgress(SoftwareUpdate, "Default Provider timer handler is invoked");
+
+    // Determine which provider to query next
+    ProviderLocation::Type providerLocation;
+    if (DetermineProviderLocation(providerLocation) != true)
+    {
+        StartDefaultProviderTimer();
+        return;
+    }
+
+    mRequestor->SetCurrentProviderLocation(providerLocation);
+    mLastUsedProvider = providerLocation;
+
+    DriverSendQuery();
+}
+
+void GenericOTARequestorDriver::StartDefaultProviderTimer()
+{
+    ChipLogProgress(SoftwareUpdate, "Starting the Default Provider timer, timeout: %u seconds",
+                    (unsigned int) mDefaultProviderTimeoutSec);
+
+    ScheduleDelayedAction(
+        System::Clock::Seconds32(mDefaultProviderTimeoutSec),
+        [](System::Layer *, void * context) {
+            (static_cast<GenericOTARequestorDriver *>(context))->DefaultProviderTimerHandler(nullptr, context);
+        },
+        this);
+}
+
+void GenericOTARequestorDriver::StopDefaultProviderTimer()
+{
+    ChipLogProgress(SoftwareUpdate, "Stopping the Default Provider timer");
+    CancelDelayedAction(
+        [](System::Layer *, void * context) {
+            (static_cast<GenericOTARequestorDriver *>(context))->DefaultProviderTimerHandler(nullptr, context);
+        },
+        this);
+}
+
+// Determines the next available Provider location and sets it in the OTARequestor
+void GenericOTARequestorDriver::DetermineAndSetProviderLocation()
+{
+    ProviderLocation::Type providerLocation;
+    if (DetermineProviderLocation(providerLocation) != true)
+    {
+        return;
+    }
+
+    mRequestor->SetCurrentProviderLocation(providerLocation);
+    mLastUsedProvider = providerLocation;
+}
+
+// Returns the next available Provider location
+bool GenericOTARequestorDriver::DetermineProviderLocation(ProviderLocation::Type & providerLocation)
```

Comment:
I think we can just combine this function with `DetermineAndSetProviderLocation`. I can't think of a reason for us to find a provider location to use but not set it.

Reply:
The reason could be that we want to take additional info into the account (like the last provider we used) before setting it. That's how this function is used in this PR.

Reply:
If we are using the last provider (or any specific provider), then we wouldn't bother calling this function

Reply:
I see your usage below where UpdateNotAvailable occurs. You are using this as a way to determine if you should try again.